narwhal/consensus/src/bullshark.rs

```diff
@@ -225,30 +234,34 @@ impl Bullshark {
         if leader_round <= state.last_round.committed_round {
             return Ok((Outcome::LeaderBelowCommitRound, Vec::new()));
         }
-        let (leader_digest, leader) = match Self::leader(&self.committee, leader_round, &state.dag)
-        {
-            Some(x) => x,
-            None => {
-                self.last_leader_election = LastRound {
-                    leader_found: false,
-                    leader_has_support: false,
-                };
-                // leader has not been found - we don't have any certificate
-                return Ok((Outcome::LeaderNotFound, Vec::new()));
-            }
-        };
+
+        let (leader, leader_authority) =
+            match Self::leader(&self.committee, leader_round, &state.dag) {
+                (Some(certificate), leader_authority) => (certificate, leader_authority),
```

Comment:
It seems more straightforward to calculate leader_authority first in a separate function, then use it to get leader certificate from dag separately.